Add support for download reproducers by TestRun ID

katieworton opened the pull request:
Add support in squadutilslib and squad-create-reproducer for downloading the build or test reproducer by TestRun ID as an alternative to searching for the latest test reproducer in a given group and project.

katieworton committed:
squadutilslib: Add get_reproducer_from_testrun

Add get_reproducer_from_testrun to allow downloading and returning of a
reproducer from a given testrun ID. This function downloads the
reproducer for the given testrun to a provided filename. The optional
`local` flag can be provided to download the tuxmake or tuxrun
reproducer command rather than the tuxsuite reproducer command.

This function will be used in a script to download the reproducer file
for a build or test from a TestRun, which is useful for reproducing
builds and tests.

Update get_reproducer to use the get_reproducer_from_testrun function to
download the reproducer from the testrun.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squad-create-reproducer-from-testrun: Download reproducers by testrun

Add support for downloading reproducers by testrun ID in the
squad-create-reproducers script. This allows the user to provide the
--testrun parameter to download the build or test reproducer from a
specific testrun, using the get_reproducer_from_testrun function from
squadutilslib.

The --local flag can be provided to fetch TuxRun reproducer, otherwise
the TuxPlan reproducer will be downloaded by default.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton commented:
@roxell if you are able to look over this PR when you have the opportunity that would be great.

I changed the input from the planned "job_url" input to the testrun ID since I discovered downloading the build reproducer requires the metadata file (which is accessible via the testrun).

roxell commented:
This should probably be in its own script with its own parameters and then just call the common parts in squadutilslib to make it easier since the imput variables are so different or what do you think?

katieworton commented:
@roxell I have created a new script `squad-create-reproducer-from-testrun` and reverted the old script.

I also updated `get_reproducer` in `squadutilslib.py` to use the `get_reproducer_from_testrun` utility to do its download step.

Let me know what you think.

Also - I noticed the flake8 check only checks the `.py` files of the repo. I am not sure if it was intentionally set up this way, but I had some unused imports that it missed as a result. If we _do_ want to run flake8 on the other Python files perhaps something like `flake8 --exclude=dist/ --ignore=E501,W503 . squad-*` could be used.

roxell commented on squad-create-reproducer:
```diff
@@ -10,7 +10,6 @@
 from logging import INFO, basicConfig, getLogger
 from os import chmod, getenv
-from pathlib import Path
 from stat import S_IRUSR, S_IWUSR, S_IXUSR
```
Please add this into the commit message for this patch

'Fixes: c6c44affc82b ("Change reproducer functions to return string rather than file name")'

katieworton replied:
Done

roxell commented on squadutilslib.py:
```diff
@@ -193,20 +237,11 @@ def get_reproducer(
         return (
-            Path(reproducer).read_text(encoding="utf-8"),
+            reproducer,
             Build(getid(testrun.build)).metadata.git_describe,
```
all the changes in this file can go into the same commit.

katieworton replied:
Done

katieworton commented:
@roxell I have made the suggested updates and created a PR for fixing flake8 checks https://github.com/Linaro/squad-client-utils/pull/34

katieworton commented:
Latest pushes add support for fetching tux plan reproducers. Currently, the "test" side does not work as the new TuxSuite patch does not appear to expose the download URL for the tux plan test reproducer in the same way as the other test reproducer URLs. I have followed up on this with the Tux team.

katieworton commented:
The Tux team followed up by added an endpoint to download the tux plan reproducers for tests. I have updated my patch to perform this download and have tested that it works.

@roxell I have added support for fetching tux plan reproducers - if you could have a look over this patch again that would be great.

roxell commented:
When I ran the command "./squad-create-reproducer-from-testrun --testrun 21531365 --plan --filename 21531365.json" I didn't expect to see any output in the console. 
I mean when --filename are passed no need to print the output in the console then.

katieworton commented:
@roxell - I have now pushed an update which means the reproducer is only printed when `--filename` is not provided.

roxell commented:
we do not need the yaml file to be exeutable...
maybe we just need to provide the url to the plan_file since tuxsuite can handle a url too what's your thoughts?

katieworton commented:
This is true - I think in both cases my only concern would be the number of code paths to maintain.

katieworton commented:
I have removed the feature of fetching TuxTest reproducers to improve maintainability (since it is unnecessary if we download TuxPlans) and updated things so plans don't get marked as executable. I have updated the commit message and README to reflect these changes.

@roxell let me know what you think of these changes :+1:

roxell commented on squad-create-reproducer-from-testrun:
```diff
@@ -0,0 +1,99 @@
+        filename = args.filename
+
+    try:
+        reproducer, is_test_reproducer = get_reproducer_from_testrun(
```
when is is_test_reproducer used?

Can't that be dropped?

katieworton replied:
Have now dropped this 👍
